ivy/functional/frontends/jax/numpy/mathematical_functions.py

```diff
@@ -624,6 +624,31 @@ def polyint(p, m=1, k=None):
     return ivy.divide(ivy.concat((p, k_arr)), coeff).astype(p.dtype)
 
 
+@with_unsupported_dtypes(
+    {"0.3.14 and below": ("float16",)},
+    "jax",
+)
+@to_ivy_arrays_and_back
+def polydiv(u, v, *, trim_leading_zeros=False):
+    dtypeu = u.dtype
+    dtypev = v.dtype
+    u = ivy.asarray(u)
+    v = ivy.asarray(v)
+    m = u.size - 1
+    n = v.size - 1
+    scale = 1. / v[0]
+    q = ivy.zeros(max(m - n + 1, 1), dtype=u.dtype)
+    for k in range(0, m - n + 1):
+        d = scale * u[k]
+        q[k] = d
+        u[k:k+n+1] += (-d * v)
+    if trim_leading_zeros:
+        u = trim_zeros(u, trim='f')
+    q = ivy.asarray(q, dtype=dtypev)
+    u = ivy.asarray(u, dtype=dtypeu)
+    return q, u
+
```

Comment:
There is also an assertion error in the CI checks indicating something is not right in the polydiv calculation. Please look into that. I would suggest running the test locally as it will help make debugging faster.

Reply:
Yes, I thought it is related to tolerance or dtypes(`[1] != [-0.0]`). I am working on it.

Reply:
I couldn't find `polydiv` at other backends. I checked again and can't find any problem with logic. However, I need to check which functions be called from different frameworks so that I can tell the problem. (their results are in different formats 
```py
import numpy as np
import tensorflow as tf
import torch as pt

# my implementation
def polydiv(u, v, *, trim_leading_zeros=False):
    m = len(u) - 1
    n = len(v) - 1
    scale = 1. / v[0]
    q = np.zeros((max(m - n + 1, 1),), dtype=u.dtype)
    for k in range(0, m - n + 1):
        d = scale * u[k]
        q[k] = d
        u[k:k+n+1] -= (d * v)
    if trim_leading_zeros:
        u = np.trim_zeros(u, trim='f')
    return q, u


x = np.array([-1.0,-1.0], dtype='float64')
y = np.array([-1.0,-1.0], dtype='float64')

a = pt.tensor([-1.0,-1.0])
b = pt.tensor([-1.0,-1.0])

print(polydiv(x, y)) # quotient, remainder
print(pt.div(a,b)) # quotient
tf.divide(x,y) # remainder
```
Result from above code:
```
(array([1.]), array([0., 0.]))
tensor([1., 1.])
tf.Tensor([-0. -0.], shape=(2,), dtype=float64)
```

Reply:
There is no point in comparing results of functions in other frameworks. This is a frontend function. It should only be compared to the corresponding original one, in this case `jax.numpy.polydiv`.

Reply:
it's weird. If I didn't totally misunderstand the `polydiv` function, jax returning wrong result. I wasn't able to install jax but tried it on google colab and it seems it returns wrong result.

Reference:
https://numpy.org/doc/stable/reference/generated/numpy.polydiv.html

Our failure:

(-x-1 ) / (-x-1) 

This is 1. Am I missing something?


![firefox_j9iAqx8rwD](https://github.com/unifyai/ivy/assets/51514616/e77c2cc6-3f25-4997-a760-fb0c0c6eeb78)

Reply:
I too would expect the result to be 1.0 with 0 remainder. This is strange.

Reply:
I would suggest the following:
1. Check if `numpy.polydiv` returns the same result. `jax.numpy.polydiv` is supposed to mimic `numpy.polydiv` , so if the result is different, it would be a hint that there is a bug in jax's source code.
2. Copy-paste jax's (and maybe numpy's) source code and run the debugger step-by-step through those functions and your function. This way we can investigate where the different result stems from.

Reply:
`numpy.polydiv` returns zero too as quotient.  I just submitted a bug report at numpy repository (Probably won't make any difference for us.). I'll debug it but even if we find the problem, how can we proceed? Isn't our implementation the one which returns correct result?

Reply:
Let me try to explain what I did. First of all I realized the code I tried on colab has a problem. When I return `u` with my `polydiv` implementation, it changes the original input and that's why `np.poydiv` or `jax.polydiv` returns wrong result if I run them with my implementation.
However, something similiar happens inside our tests.  Now I'll explain why I think that. This is my implementation:
```py
@with_unsupported_dtypes(
    {"0.3.14 and below": ("float16",)},
    "jax",
)
@to_ivy_arrays_and_back
def polydiv(u, v, *, trim_leading_zeros=False):
#    if (u == ivy.array([-1.])).all() and (v == ivy.array([-1.])).all():
#        return [ivy.array([-0.], dtype=u.dtype), ivy.array([0.], dtype=u.dtype)]
#    if (u == ivy.array([-1., -1.])).all() and (v == ivy.array([-1., -1.])).all():
#        return [ivy.array([-0.], dtype=u.dtype), ivy.array([0., 0.], dtype=u.dtype)]
    u_arr, v_arr = ivy.promote_types_of_inputs(u, v)
    m = len(u_arr) - 1
    n = len(v_arr) - 1
    scale = 1. / v_arr[0]
    q = ivy.zeros((max(m - n + 1, 1),))
    for k in range(0, m - n + 1):
        d = scale * u_arr[k]
        q[k] = d
        u_arr[k:k+n+1] -= d * v_arr
    if trim_leading_zeros:
        u_arr = ivy.trim_zeros(u_arr, trim='f')
    return q, u_arr
```
I just didn't change input arrays so I can exclude that failure. However this didn't work.
input for u and v:
```
E            [array([-1.], dtype=float32), array([-1.], dtype=float32)]),
```
and the error:
```
E       len(ret_np_flat) != len(ret_np_from_gt_flat):
E       
E       ret_np_flat:
E       
E       [array([1.])]
E       
E       ret_np_from_gt_flat:
E       
E       [array([-0.], dtype=float32), array([0.], dtype=float32)]
```

If I am not wrong, our implementation returns 1 as expected. Than I wrote the part commented out at above. If we run test with removing `#` for first if block:
```
E           AssertionError:  the results from backend numpy and ground truth framework jax do not match
E            [-0.]!=[1.] 
```
Now test complains our result is wrong.


Also, I tried to vary tolerances and min/max values. Moreover, I debugged by breakpoints and used `print` s for logging. 

```py
x = 0
@with_unsupported_dtypes(
    {"0.3.14 and below": ("float16",)},
    "jax",
)
@to_ivy_arrays_and_back
def polydiv(u, v, *, trim_leading_zeros=False):
    global x
    x += 1
    print(x)
    print(f"input arrays u: {u} and v: {v}")
    u_arr, v_arr = ivy.promote_types_of_inputs(u, v)
    m = len(u_arr) - 1
    n = len(v_arr) - 1
    print(f"promoted input arrays u_arr: {u_arr} and v_arr: {v_arr}")
    scale = 1. / v_arr[0]
    q = ivy.zeros((max(m - n + 1, 1),), dtype=u_arr.dtype)
    for k in range(0, m - n + 1):
        d = scale * u_arr[k]
        q[k] = d
        u_arr[k:k+n+1] -= (d * v_arr)
    if trim_leading_zeros:
        u_arr = trim_zeros(u_arr, trim='f')
    print(f"output arrays q: {q} and u_arr {u_arr}")
    return q, u_arr
```

This shows me that my function works correct.  Logs are below from numpy:
```
1
input arrays u: ivy.array([-1.]) and v: ivy.array([-1.])
input arrays u_arr: [-1.] and v_arr: [-1.]
output arrays q: ivy.array([1.]) and u_arr [0.]
2
input arrays u: ivy.array([-1.]) and v: ivy.array([-1.])
input arrays u_arr: [-1.] and v_arr: [-1.]
output arrays q: ivy.array([1.]) and u_arr [0.]
3
input arrays u: ivy.array([-1.]) and v: ivy.array([-1.])
input arrays u_arr: [-1.] and v_arr: [-1.]
output arrays q: ivy.array([1.]) and u_arr [0.]
4
input arrays u: ivy.array([-1.]) and v: ivy.array([-2.])
input arrays u_arr: [-1.] and v_arr: [-2.]
output arrays q: ivy.array([0.5]) and u_arr [0.]
5
input arrays u: ivy.array([-1.]) and v: ivy.array([-2.])
input arrays u_arr: [-1.] and v_arr: [-2.]
output arrays q: ivy.array([0.5]) and u_arr [0.]
6
input arrays u: ivy.array([-1.]) and v: ivy.array([-2.])
input arrays u_arr: [-1.] and v_arr: [-2.]
output arrays q: ivy.array([0.5]) and u_arr [0.]
7
input arrays u: ivy.array([-1.]) and v: ivy.array([-2.11664224e-06])
input arrays u_arr: [-1.] and v_arr: [-2.1166422e-06]
output arrays q: ivy.array([472446.4062]) and u_arr [0.]
8
input arrays u: ivy.array([-1.]) and v: ivy.array([-1.])
input arrays u_arr: [-1.] and v_arr: [-1.]
output arrays q: ivy.array([1.]) and u_arr [0.]
9
input arrays u: ivy.array([-1.]) and v: ivy.array([-257.])
input arrays u_arr: [-1.] and v_arr: [-257.]
output arrays q: ivy.array([0.00389105]) and u_arr [0.]
10
input arrays u: ivy.array([-257.]) and v: ivy.array([-1.])
input arrays u_arr: [-257.] and v_arr: [-1.]
output arrays q: ivy.array([257.]) and u_arr [0.]
11
input arrays u: ivy.array([-1.]) and v: ivy.array([-1.])
input arrays u_arr: [-1.] and v_arr: [-1.]
output arrays q: ivy.array([1.]) and u_arr [0.]
12
input arrays u: ivy.array([-1.]) and v: ivy.array([1.])
input arrays u_arr: [-1.] and v_arr: [1.]
output arrays q: ivy.array([-1.]) and u_arr [0.]
13
input arrays u: ivy.array([-1.]) and v: ivy.array([-1.])
input arrays u_arr: [-1.] and v_arr: [-1.]
output arrays q: ivy.array([1.]) and u_arr [0.]
14
input arrays u: ivy.array([-1.]) and v: ivy.array([-1.])
input arrays u_arr: [-1.] and v_arr: [-1.]
output arrays q: ivy.array([1.]) and u_arr [0.]
15
input arrays u: ivy.array([-1.]) and v: ivy.array([-1.])
input arrays u_arr: [-1.] and v_arr: [-1.]
output arrays q: ivy.array([1.]) and u_arr [0.]
```

Lastly, I think problem is related to the loop I used. Below part can't return anything from `jax` tests. The original implementation of the function uses these:
https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html

I changed that block:
```py
  for k in range(0, m-n+1):
    d = scale * u_arr[k]
    q = q.at[k].set(d)
    u_arr = u_arr.at[k:k+n+1].add(-d*v_arr)
```
to
```py
    for k in range(0, m - n + 1):
        d = scale * u_arr[k]
        q[k] = d
        u_arr[k:k+n+1] -= (d * v_arr)
```

`jax` part of tests returns the `TypeError` below:
```
>     raise TypeError(msg.format(type(self)))
E     TypeError: '<class 'jaxlib.xla_extension.ArrayImpl'>' object does not support item assignment. JAX arrays are immutable. Instead of ``x[idx] = y``, use ``x = x.at[idx].set(y)`` or another .at[] method: https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html
```

Both of these two approaches do the same thing however, jax arrays are immutable so indexing doesn't work.
For reference, original implementation:
```py
@_wraps(np.polydiv, lax_description=_LEADING_ZEROS_DOC)
def polydiv(u: ArrayLike, v: ArrayLike, *, trim_leading_zeros: bool = False) -> Tuple[Array, Array]:
  check_arraylike("polydiv", u, v)
  u_arr, v_arr = promote_dtypes_inexact(u, v)
  m = len(u_arr) - 1
  n = len(v_arr) - 1
  scale = 1. / v_arr[0]
  q: Array = zeros(max(m - n + 1, 1), dtype = u_arr.dtype) # force same dtype
  for k in range(0, m-n+1):
    d = scale * u_arr[k]
    q = q.at[k].set(d)
    u_arr = u_arr.at[k:k+n+1].add(-d*v_arr)
  if trim_leading_zeros:
    # use the square root of finfo(dtype) to approximate the absolute tolerance used in numpy
    u_arr = trim_zeros_tol(u_arr, tol=sqrt(finfo(u_arr.dtype).eps), trim='f')
  return q, u_arr
```

Note: I am not testing `trim_leading_zeros` just to investigate problem with a simpler aspect.

Reply:
Thanks for explaining in detail!
So, there is the error that the ground truth returns 0 instead of 1 but when you manually set the code to return 0 in that case there is an error that the ground truth returns 1? 😅 Could this be related to rounding depending on the data type?
About the item assignment error, in which function does this occur? I guess `u_arr` is an ivy array, so the error must be coming up in the jax backend of some ivy function.
Also why do you think the for loop is in blame for the incorrect returned values? Did the assertion error get fixed for the rest of the backends when you changed that assignment line?

Reply:
> Thanks for explaining in detail! So, there is the error that the ground truth returns 0 instead of 1, but when you manually set the code to return 0 in that case the error says that the ground truth returns 1? 😅 Could this be related to rounding depending on the data type? 

- Exactly. Since I thought this could be a problem, I played with min/max values and tolerances. Varied them to both edges ( decresed and increased tolerances/min, max values). I am not sure if this enough to eliminate that possibility.

> About the item assignment error, in which function does this occur? I guess `u_arr` is an ivy array, so the error must be coming up in the jax backend of some ivy function.

- While implementing those functions I usually fix errors from `numpy` parts without checking rest of tests. And once I solve errors on there, all tests are usually passing. For this case, I just realized `jax` part fails at each case because of the reason explained below. I don't know which Ivy function throws the error. Probably some fancy wrapper on there that I have no idea what's happening behind. 🤣 Here's the stack info from  the error:
```
test_jax_numpy_math.py:2775: in test_jax_numpy_polydiv
    helpers.test_frontend_function(
../../helpers/function_testing.py:671: in test_frontend_function
    ret = get_frontend_ret(
../../helpers/function_testing.py:2002: in get_frontend_ret
    ret = frontend_fn(*args, **kwargs)
/ivy/ivy/functional/frontends/jax/func_wrapper.py:102: in _outputs_to_frontend_arrays_jax
    ret = fn(*args, **kwargs)
/ivy/ivy/functional/frontends/jax/func_wrapper.py:79: in _inputs_to_ivy_arrays_jax
    return fn(*new_args, **new_kwargs)
/ivy/ivy/functional/frontends/jax/numpy/mathematical_functions.py:645: in polydiv
    u_arr[k:k+n+1] -= (d * v_arr)
```

> Also why do you think the for loop is in blame for the incorrect returned values? Did the assertion error get fixed for the rest of the backends when you changed that assignment line?

- Because jax part of the tests never outputs with my logging implementation.(Logs below)  At item assignment part they throw the error I mentioned. Jax arrays are immutable. Jax implementation of the `polydiv` function uses that approach:
https://jax.readthedocs.io/en/latest/_autosummary/jax.numpy.ndarray.at.html
I also imported jax and used `at` inside the loop but than our function returns nothing. (empty array)
So, to solve this, I probably need to use inplace updates but don't know where to look at. As far as I remember our framework also has that feature, I'll check docs when I am able to.
```
input arrays u: ivy.array([-1.]) and v: ivy.array([-1.])
promoted input arrays u_arr: [-1.] and v_arr: [-1.]
17
input arrays u: ivy.array([-1.]) and v: ivy.array([-1.])
promoted input arrays u_arr: [-1.] and v_arr: [-1.]
18
input arrays u: ivy.array([-1.]) and v: ivy.array([-1.])
promoted input arrays u_arr: [-1.] and v_arr: [-1.]
19
.
.
.
```

If you have anything that can lead me to a right direction, I can follow. Thanks!